server/mqtt.go

```diff
@@ -4142,6 +4142,41 @@ func mqttSubscribeTrace(pi uint16, filters []*mqttFilter) string {
 	return sb.String()
 }
 
+// mqttDoNotDeliverSubject returns true if the subject starts with $JS or $MQTT
+// since we don't want to deliver them, avoiding session deadlocks with JSAPI
+// invocations.
+func mqttDoNotDeliverSubject(subject string) bool {
+	l := len(subject)
+	if l == 0 {
+		return true
+	}
+	if subject[0] != '$' {
+		return false
+	}
+	if l < 4 {
+		return false
+	}
+	if subject[1] != 'J' && subject[1] != 'M' {
+		return false
+	}
+	if subject[2] != 'S' && subject[1] != 'Q' {
+		return false
+	}
+	if subject[3] != '.' && subject[3] != 'T' {
+		return false
+	}
+	if l < 6 {
+		return false
+	}
+	if subject[4] != 'T' {
+		return false
+	}
+	if subject[5] != '.' {
```

Comment:
Not just `$MQTT.` because otherwise I think you will fail some QoS2 tests. It has to be the `mqttJSARepliesPrefix`.

Reply:
Right.

Reply:
It wasn't even right to begin with, added a test.